### PR TITLE
Fix fetch quay

### PR DESCRIPTION
**Pull request by Midnighter**
I was in the mood for writing something pretty so I went a bit overboard with this fix. Changes:

* Follow `next_page`s
* Instead of `skip.list` being regular expressions, check for image names starting with the given strings
* Log more information
* Rich progress report
* Use asynchronous requests for concurrency, can be tweaked to respect API rate limits
* Use modern Python syntax

### CODE REVIEW

**Comment by drpatelh** [comment]
Awesome! Thanks @Midnighter 🚀

**Comment by Midnighter** [comment]
@drpatelh recommended, it might be helpful to post script output so here goes:

```bash
[22:14:13] INFO     Fetching quay.io BioContainers images.            get_container_list.py:374
Image Batch ⠸ 99/?   0:00:28
Image Tags ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9,880/9,880(100.0%)   0:00:00   0:16:38
[22:31:21] INFO     Found 77,010 images with tags.                    get_container_list.py:376
           INFO     Fetching Singularity BioContainers images.        get_container_list.py:380
           WARNING  No images found at 'https://depot.galaxyproject.o get_container_list.py:252
                    rg/singularity/new/'.
[22:31:26] INFO     Found 71,296 images with tags.                    get_container_list.py:384
           INFO     Parsing container image deny list.                get_container_list.py:388
           INFO     2,646 new images found. Generating build script.  get_container_list.py:394
```

What you can see is that it goes through all the pages with image repositories on quay.io in about 30 seconds. Then it fetches all the tags for the almost 10 k images in about 16 min. Their combination yields close to 77 k images.

Then 71 k singularity images are found. Their difference and applying the skip list yields a final 2,600 to be built. Seems like quite a lot but I don't know for how long the bot hasn't been working properly.

**Review by simonbray — COMMENTED** [review]
Very nice, thanks @Midnighter!

Do you have any idea what is causing those missing containers? We never noticed any problems until very recently.

**Comment by Midnighter** [comment]
> Do you have any idea what is causing those missing containers? We never noticed any problems until very recently.

Not yet, no. Could pick a few random ones and check manually?

File: get_container_list.py [hunk 1]
```diff
@@ -1,91 +1,399 @@
...
+        wait=tenacity.wait_random_exponential(),
+        stop=tenacity.stop_after_attempt(5),
+        reraise=True,
+        retry=tenacity.retry_if_exception_type(httpx.HTTPError),
+        before=tenacity.before_log(logger, logging.DEBUG),
+    )
+    def _fetch_images(client: httpx.Client, url: str) -> Optional[str]:
+        """Make a single GET request and return the response body as text."""
+        response = client.get(url=url)
+        # FIXME: It seems that 404 indicates no new images. Or is `new` an outdated path?
```
[the remark this answers is not in the document]
Comment (Midnighter) [reply get_container_list.py:266]: Cool, I'll remove the URL.

File: get_container_list.py [hunk 2]
```diff
@@ -1,91 +1,399 @@
...
+def get_new_images(
+    quay_images: Iterable[str],
+    singularity_images: Iterable[str],
+    denylist: Iterable[str],
+) -> List[str]:
+    """Identify new images from the given lists."""
+    denylist = tuple(denylist)
+    result = frozenset(quay_images) - frozenset(singularity_images)
+    # Filter new images using the deny list.
+    # FIXME: Is it necessary to sort bioconductor images to the end as before?
```
[the remark this answers is not in the document]
Comment (Midnighter) [reply get_container_list.py:282]: You got it.

**Comment by simonbray** [comment]
Okay, let's try this out. Thanks again @Midnighter!

**Comment by Midnighter** [comment]
Doh, slightly too soon with the merge. There was a mistake.

**Pull request merged**